ascon-aead/src/asconcore.rs

```diff
@@ -0,0 +1,379 @@
+            self.state[sidx] ^= pad(associated_data.len());
+            if !associated_data.is_empty() {
+                self.state[sidx] ^= u64_from_be_bytes_partial(associated_data);
+            }
+            self.permute_state();
+        }
+
+        // domain separation
+        self.state[4] ^= 1;
+    }
+
+    fn process_encrypt_inplace(&mut self, mut message: &mut [u8]) {
+        while message.len() >= P::COUNT {
+            // process full block of message
+            self.state[0] ^= u64_from_be_bytes(&message[..8]);
+            message[..8].copy_from_slice(&u64::to_be_bytes(self.state[0]));
+            if P::COUNT == 16 {
+                self.state[1] ^= u64_from_be_bytes(&message[8..16]);
+                message[8..16].copy_from_slice(&u64::to_be_bytes(self.state[1]));
+            }
+            self.permute_state();
+            message = &mut message[P::COUNT..];
+        }
+
+        // process partial block if it exists
+        let sidx = if P::COUNT == 16 && message.len() >= 8 {
+            self.state[0] ^= u64_from_be_bytes(&message[..8]);
+            message[..8].copy_from_slice(&u64::to_be_bytes(self.state[0]));
+            message = &mut message[8..];
+            1
+        } else {
+            0
+        };
+        self.state[sidx] ^= pad(message.len());
+        if !message.is_empty() {
+            self.state[sidx] ^= u64_from_be_bytes_partial(message);
+            message.copy_from_slice(&u64::to_be_bytes(self.state[sidx])[0..message.len()]);
+        }
+    }
+
+    fn process_decrypt_inplace(&mut self, mut ciphertext: &mut [u8]) {
+        while ciphertext.len() >= P::COUNT {
+            // process full block of ciphertext
+            let cx = u64_from_be_bytes(&ciphertext[..8]);
+            ciphertext[..8].copy_from_slice(&u64::to_be_bytes(self.state[0] ^ cx));
+            self.state[0] = cx;
+            if P::COUNT == 16 {
+                let cx = u64_from_be_bytes(&ciphertext[8..16]);
+                ciphertext[8..16].copy_from_slice(&u64::to_be_bytes(self.state[1] ^ cx));
+                self.state[1] = cx;
+            }
+            self.permute_state();
+            ciphertext = &mut ciphertext[P::COUNT..];
+        }
+
+        // process partial block if it exists
+        let sidx = if P::COUNT == 16 && ciphertext.len() >= 8 {
+            let cx = u64_from_be_bytes(&ciphertext[..8]);
+            ciphertext[..8].copy_from_slice(&u64::to_be_bytes(self.state[0] ^ cx));
+            self.state[0] = cx;
+            ciphertext = &mut ciphertext[8..];
+            1
+        } else {
+            0
+        };
+        self.state[sidx] ^= pad(ciphertext.len());
+        if !ciphertext.is_empty() {
+            let cx = u64_from_be_bytes_partial(ciphertext);
+            self.state[sidx] ^= cx;
+            ciphertext.copy_from_slice(&u64::to_be_bytes(self.state[sidx])[0..ciphertext.len()]);
+            self.state[sidx] = clear(self.state[sidx], ciphertext.len()) ^ cx;
+        }
```

Comment:
This seems like a candidate for `chunks_exact_mut`

Reply:
Changed to use `chunks_exact_mut`.